File: src/pattern/iterator/java/BookShelf.java
package pattern.iterator.java;

public class BookShelf implements Aggregate {
    private int last = 0;
    private Book[] books;

    public BookShelf(int maxLength){
        books = new Book[maxLength];
    }

    public Book getBookAt(int index){
        return books[index];
    }

    public int getLength(){
        return this.books.length;
    }

    public void appendBook(Book book){
        this.books[last] = book;
        last++;
    }

    public Iterator iterator(){
        return new BookShelfIterator(this);
    }

}

File: src/pattern/iterator/kotlin/BookShelf.kt
package pattern.iterator.kotlin

import java.lang.reflect.Array

class BookShelf(maxLength: Int): Aggregate{


    var last: Int = 0
//    val books = Array<Book>(maxLength, { Book("KIM") })
    var books = arrayListOf<Book>()

    open fun getBookAt(index: Int): Book{
//        return this.books[index]
        return this.books.get(index)
    }

    open fun appendBook(book: Book): Unit{
        //this.books[this.last] = book
        this.books.add(book)
        this.last++
    }

    open fun getLength(): Int{
//        return this.books.size
        return this.books.size
    }


    override fun iterator(): Iterator{
        return BookShelfIterator(0,this)
    }

}